dimod/higherorder/polynomial.py

```diff
@@ -34,6 +37,22 @@ def asfrozenset(term):
     return term if isinstance(term, frozenset) else frozenset(term)
 
 
+def freeze_term(term, vartype):
+    return (
+        term
+        if isinstance(term, frozenset)
+        else (
+                frozenset(term)
+                if vartype is Vartype.BINARY
+                else frozenset(
+                        var
+                        for var, power in tl.frequencies(term).items()
```

Comment:
Seems like a big dependency to bring in for a [4 line function](https://github.com/pytoolz/toolz/blob/e9bc1e150b7b1ea1bec300a55382ed564b215f41/toolz/itertoolz.py#L548).

Reply:
I also suspect that in typical use cases, i.e. relatively low degree terms, that just doing `tuple.count()` will be faster than constructing a new `defaultdict`.

Reply:
```python
In [7]: term = tuple('abcde')

In [8]: %timeit for _ in frequencies(term): pass
1.31 µs ± 9.02 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [9]: %timeit for _ in ((v, term.count(v)) for v in term): pass
979 ns ± 7.79 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
I also wonder if [Counter](https://docs.python.org/3/library/collections.html#collections.Counter) would do the same thing.

Reply:
That result seems to be from the interpreted `toolz` version of frequencies. The compiled cytools version is >2x faster than the `.count` version. The count version need to iterate over `set(terms)` rather than term which slows it down more, making it closer to 3x slower than compiled cytoolz.
Counter is around the same speed as the interpreted version of `frequencies`.

```python
%timeit for _ in toolz.frequencies(term).items(): pass
    1.16 µs ± 23.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit for _ in ((v, term.count(v)) for v in term): pass
    826 ns ± 46.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit for _ in cytoolz.frequencies(term).items(): pass
    347 ns ± 10.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit for _ in ((v, term.count(v)) for v in set(term)): pass
    1.01 µs ± 49.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit for _ in ((v, count) for v, count in Counter(term).items()): pass
    1.46 µs ± 15.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit for _ in Counter(term).items(): pass
    1.08 µs ± 20.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
> 
> 
> Seems like a big dependency to bring in for a [4 line function](https://github.com/pytoolz/toolz/blob/e9bc1e150b7b1ea1bec300a55382ed564b215f41/toolz/itertoolz.py#L548).

Cytoolz offers 2 benefits: clean code and speed.
The speed it brings is often sufficient to avoid needing to cythonize or port code to C leading to further code quality benefits as well
Depending on it brings in a moderately large number of small utility functions only a small number of which is likely to be used in any given PR. In each case it is easy to either reimplement an equivalent function or find a work around. In the long run though that means either having most of the library reimplemented possibly multiple times throughout our codebase and/or a lot of workarounds. Individual reimplementations are unlikely to get cythonized, so the codebase misses out on the performance benefit, in some cases leading to unnecessary C ports.

This could by demonstrated by refactoring a large piece of  code that is performance bottleneck. I will do more refactoring and optimization in this file for now to see how many more cytoolz uses turn up. So far i have been doing it method by method to keep the PRs small but a combine PR would demonstrate the cumulative effect